File: python/leetcode/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        start_n = 0
        end_n = len(matrix) - 1
        start_m = 0
        end_m = len(matrix[0]) - 1 if len(matrix) > 0 else - 1
        result = []
        
        while start_n <= end_n and start_m <= end_m:
            # Go across the top of the matrix
            for i in range(start_m, end_m + 1):
                result.append(matrix[start_n][i])
            start_n += 1
        
            # Go down the right of the matrix
            for i in range(start_n, end_n + 1):
                result.append(matrix[i][end_m])
            end_m -= 1

            # Go back along the bottom
            if start_n <= end_n:
                for i in range(end_m, start_m - 1, -1):
                    result.append(matrix[end_n][i])
                end_n -= 1

            # Go up along the left
            if start_m <= end_m:
                for i in range(end_n, start_n - 1, -1):
                    print(i)
                    result.append(matrix[i][start_m])
                start_m += 1
        return result

File: java/leetcode/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int startN = 0;
        int endN = matrix.length - 1;
        int startM = 0;
        int endM = matrix.length > 0 ? matrix[0].length - 1 : -1;
        int[] result = new int[(endN + 1) * (endM + 1)];
        int idx = 0;
        while (startN <= endN && startM <= endM) {
            for (int i = startM; i <= endM; i++) {
                result[idx] = matrix[startN][i];
                idx++;
            }
            startN++;

            for (int i = startN; i <= endN; i++) {
                result[idx] = matrix[i][endM];
                idx++;
            }
            endM--;

            if (startN <= endN) {
                for (int i = endM; i >= startM; i--) {
                    result[idx] = matrix[endN][i];
                    idx++;
                }
                endN--;
            }

            if (startM <= endM) {
                for (int i = endN; i >= startN; i--) {
                    result[idx] = matrix[i][startM];
                    idx++;
                }
                startM++;    
            }

        }
        return Arrays.stream(result).boxed().collect(Collectors.toList());
    }
}
